SSA/Core/Framework.lean

```diff
@@ -220,282 +590,717 @@ args, since there now are equation lemmas for it.
-    | last =>
-      apply body.denote
-      apply e.denote
-      exact ll
-    | toSnoc v =>
-      exact e.denote ll v
+end Unfoldings
+
+/-! simp-lemmas about `denote` functions -/
+section Lemmas
+
+@[simp] lemma Com.denote_ret {eff : EffectKind} (Γ : Ctxt Ty) (x : Γ.Var t) [Monad m] :
+    (Com.ret x : Com Op Γ eff t).denote = fun Γv => return (Γv x) := by
+  funext Γv
+  simp [denote]
+
+@[simp] lemma Com.denote_lete [LawfulMonad m] {e : Expr Op Γ eff α} :
+    (Com.lete e body).denote =
+    fun Γv => (e.denote (m := m) Γv) >>= (fun v => body.denote (Γv.snoc v)) := by
+  funext Γv
+  cases eff <;> simp [denote]
+
+@[simp] lemma Com.denoteLets_lete (e : Expr Op Γ eff t) (body : Com Op _ eff u) [LawfulMonad m] :
+    (Com.lete e body).denoteLets =
+        (fun V => e.denote V >>= fun Ve => body.denoteLets (V.snoc Ve)) := by
+  funext V
+  cases eff <;> simp [denoteLets, bind_pure]
+
+@[simp] lemma Com.denoteImpure_ret [Monad m] [OpDenote Op Ty m] {Γ : Ctxt Ty} (x : Γ.Var t) :
+  (Com.ret (Op := Op) (eff := eff) x).denoteImpure = fun Γv => return (Γv x) := rfl
+
+@[simp] lemma Com.denoteImpure_body [Monad m] [OpDenote Op Ty m] {Γ : Ctxt Ty}
+    (e : Expr Op Γ eff te) (body : Com Op (Γ.snoc te) eff tbody) :
+  (Com.lete e body).denoteImpure =
+  fun Γv => e.denoteImpure Γv >>= fun x => body.denote (Γv.snoc x) := rfl
+
+@[simp] lemma Lets.denote_nil {Γ : Ctxt Ty} :
+    (Lets.nil : Lets Op Γ eff Γ).denote = (return ·) := by
+  funext; simp [denote]
+-- TODO: make `[simp]`
+lemma Lets.denote_lete {lets : Lets Op Γ_in eff Γ_out} {e : Expr Op Γ_out eff t} :
+    (lets.lete e).denote = fun V_in => (do
+      let V_out ← lets.denote V_in
+      let x ← e.denote V_out
+      return V_out.snoc x) := by
+  funext V; simp [denote]
+
+
+
+@[simp] lemma Lets.denote_addComToEnd
+    {lets : Lets Op Γ_in eff Γ_out} {com : Com Op Γ_out eff t} :
+    Lets.denote (lets.addComToEnd com) = fun V => (do
+        let Vlets ← lets.denote V
+        let Vbody ← com.denoteLets Vlets
+        return Vbody
+      ) := by
+  induction com
+  case ret => simp [Lets.denote_lete, Com.denoteLets]
+  case lete ih => simp [addComToEnd, ih, denote_lete]
+
+@[simp] lemma Com.denoteLets_ret : (.ret v : Com Op Γ eff t).denoteLets = fun V => pure V := by
+  funext V; simp [denoteLets]
+
+theorem Com.denoteLets_eq {com : Com Op Γ eff t} : com.denoteLets = com.toLets.denote := by
+  simp only [toLets]; induction com using Com.rec' <;> simp [Lets.denote_lete]
+
+end Lemmas
+
+/-!
+## `changeVars`
+Map a context homomorphism over a `Expr`/`Com`.
+That is, substitute variables.
+-/
 
 def Expr.changeVars (varsMap : Γ.Hom Γ') :
-    {ty : Ty} → (e : Expr Op Γ ty) → Expr Op Γ' ty
-  | _, ⟨op, Eq.refl _, args, regArgs⟩ => ⟨op, rfl, args.map varsMap, regArgs⟩
+    {ty : Ty} → (e : Expr Op Γ eff ty) → Expr Op Γ' eff ty
+  | _, ⟨op, sig_eq, eff_leq, args, regArgs⟩ =>
+     ⟨op, sig_eq, eff_leq, args.map varsMap, regArgs⟩
+
+-- TODO: figure out what this comment means?
```

Comment:
I'd like to know as well :)

Reply:
As the author of the comment, neither do I, so I deleted the comment :D